File: apps/java/src/test/java/chapter10/TribuoClassifierTest.java
package chapter10;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import chapter02.Features;
import chapter02.TrainTestSplit;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.tribuo.classification.dtree.CARTClassificationTrainer;
import org.tribuo.classification.ensemble.VotingCombiner;
import org.tribuo.classification.sgd.linear.LogisticRegressionTrainer;
import org.tribuo.common.tree.RandomForestTrainer;

class TribuoClassifierTest {
  private static TrainTestSplit<Features, String> twoSpeciesSplit() {
    return new TrainTestSplit<>(
        Samples.twoSpeciesX(),
        Samples.twoSpeciesNewX(),
        Samples.twoSpeciesT(),
        List.of("setosa", "virginica"));
  }

  @Test
  @DisplayName("Tribuo のロジスティック回帰とランダムフォレストも同じ関数で評価できる")
  void sameFunction() {
    List<Classifier> models =
        List.of(
            new TribuoClassifier(new LogisticRegressionTrainer()),
            TribuoClassifier.randomForest(10, TribuoClassifier.UNLIMITED, 0L));

    var scores = models.stream().map(model -> Score.evaluate(model, twoSpeciesSplit())).toList();

    assertThat(scores).containsOnly(new Score(1.0, 1.0)).hasSize(2);
  }

  @Test
  @DisplayName("学習する前に予測するとエラーになる")
  void predictBeforeFit() {
    assertThatThrownBy(
            () ->
                new TribuoClassifier(new LogisticRegressionTrainer())
                    .predict(Samples.column("花弁幅", 0.1)))
        .isInstanceOf(IllegalStateException.class)
        .hasMessage("fit で学習してから predict を呼んでください");
  }

  @Test
  @DisplayName("Tribuo のランダムフォレストは分割ごとに特徴量を絞らない決定木を受け付けない")
  void tribuoRandomForestRequiresFraction() {
    var tree = new CARTClassificationTrainer(Integer.MAX_VALUE);

    assertThatThrownBy(() -> new RandomForestTrainer<>(tree, new VotingCombiner(), 100, 0L))
        .hasMessageContaining("requires that the decision tree innerTrainer have fractional");
  }
}

File: apps/kotlin/src/test/kotlin/chapter10/TribuoClassifierTest.kt
package chapter10

import chapter02.TrainTestSplit
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import org.tribuo.classification.sgd.linear.LogisticRegressionTrainer
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class TribuoClassifierTest {
    private fun twoSpeciesSplit(): TrainTestSplit<String> {
        val (x, t) = twoSpecies()
        return TrainTestSplit(
            xTrain = x,
            xTest =
                dataFrameOf(
                    "がく片幅" to listOf(0.4, 0.4),
                    "花弁幅" to listOf(0.13, 0.83),
                ),
            tTrain = t,
            tTest = listOf("setosa", "virginica"),
        )
    }

    @Test
    fun `Tribuoのロジスティック回帰とランダムフォレストも同じ関数で評価できる`() {
        val models: List<Classifier> =
            listOf(
                TribuoClassifier(LogisticRegressionTrainer()),
                tribuoRandomForest(nEstimators = 10, maxDepth = null, seed = 0L),
            )

        val scores = models.map { evaluate(it, twoSpeciesSplit()) }

        assertEquals(List(2) { Score(train = 1.0, test = 1.0) }, scores)
    }

    @Test
    fun `学習する前に予測するとエラーになる`() {
        val error =
            assertFailsWith<IllegalStateException> {
                TribuoClassifier(LogisticRegressionTrainer()).predict(dataFrameOf("花弁幅" to listOf(0.1)))
            }

        assertEquals("fit で学習してから predict を呼んでください", error.message)
    }
}
